web/app/Models/User.php

```diff
@@ -157,133 +159,259 @@ public function unfollowUser(User $user_to_unfollow) : int
+     * @param User $user
+     * @return bool
+     */
+    public function isContact(User $user): bool
+    {
+        return (
+            Interaction::check($this, InteractionType::USER_CONTACTS, $user)
+            || Interaction::check($user, InteractionType::USER_CONTACTS, $this)
+        );
+    }
 
-        return Interaction::add($this, InteractionType::USER_CONTACT_REQUESTS, $user_to_request);
+    /**
+     * Remove a user from this user's contact list, and vice-versa.
+     * @param User $user_to_remove
+     * @return bool
+     */
+    public function removeContact(User $user_to_remove): bool
+    {
+        return(
+            Interaction::remove($this, InteractionType::USER_CONTACTS, $user_to_remove)
+            || Interaction::remove($user_to_remove, InteractionType::USER_CONTACTS, $this)
+        );
     }
 
     /**
      * Deny a request from another user to be added as a contact.
      * @param User $user_to_deny
-     * @return int|null
+     * @return bool
      */
-    public function denyContactRequest(User $user_to_deny): ?int
+    public function denyContactRequest(User $user_to_deny): bool
     {
         return Interaction::remove($user_to_deny, InteractionType::USER_CONTACT_REQUESTS, $this);
     }
 
+    /**
+     * Deny a request from another user to be added as a contact, and block them from further interaction.
+     * @param User $user_to_deny
+     * @return bool
+     */
+    public function denyContactRequestAndBlock(User $user_to_deny): bool
+    {
+        $this->blockUser($user_to_deny);
+        return Interaction::remove($user_to_deny, InteractionType::USER_CONTACT_REQUESTS, $this);
+    }
+
     /**
      * Cancel a pending request to be added as another user's contact.
      * @param User $user_to_cancel
-     * @return int|null
+     * @return bool
      */
-    public function cancelContactRequest(User $user_to_cancel): ?int
+    public function cancelContactRequest(User $user_to_cancel): bool
     {
         return Interaction::remove($this, InteractionType::USER_CONTACT_REQUESTS, $user_to_cancel);
     }
 
+    /**************************************************
+     * User Blocks User
+     *************************************************/
+
     /**
-     * Approve a request to be added as a contact of another user.
-     * @param User $user_to_approve
-     * @return Interaction|null
+     * This user blocks the target user.
+     * @param User $user_to_block
+     * @param string $reason
+     * @return bool
      */
-    public function approveContactRequest(User $user_to_approve): ?Interaction
+    public function blockUser(User $user_to_block, string $reason = '') : bool
     {
-        Interaction::remove($user_to_approve, InteractionType::USER_CONTACT_REQUESTS, $this);
-        return $this->addContact($user_to_approve);
+        /** Do not add a second block for the same target user. */
+        if($this->isBlockingUser($user_to_block)) { return false; }
```

Comment:
This makes me think, are there any checks for any interaction methods that prevents interactions with oneself? e.g. following or blocking yourself

Reply:
There are, they're in the Event controller right now, when I get to refactoring the actual controllers (requires API) we'll make some determinations on whether it makes sense to have the validation in the controller or the model method, I'm leaning towards the controller but could be convinced either way.